src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -607,13 +607,21 @@ void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
           RequestNewWorkerIfNeeded(scheduling_key);
 
         } else {
-          if (IsRayletFailed(RayConfig::instance().RAYLET_PID())) {
-            RAY_LOG(WARNING)
-                << "The worker failed to receive a response from the local "
-                   "raylet because the raylet is crashed. Terminating the worker.";
-            QuickExit();
+          if (status.IsGrpcUnavailable()) {
+            RAY_LOG(WARNING) << "The worker failed to receive a response from the local "
+                                "raylet because the raylet is unavailable (crashed).";
+            auto &task_queue = scheduling_key_entry.task_queue;
+            while (!task_queue.empty()) {
+              auto &task_spec = task_queue.front();
+              RAY_UNUSED(task_finisher_->FailOrRetryPendingTask(
+                  task_spec.TaskId(), rpc::ErrorType::SUBMITTER_NODE_DIED, &status));
+              task_queue.pop_front();
+            }
+            if (scheduling_key_entry.CanDelete()) {
+              scheduling_key_entries_.erase(scheduling_key);
+            }
```

Comment:
Is it necessary to go through the entire task queue here? It might be simpler just to mark the current task as failed, and let the rest get scheduled and failed independently.

Reply:
We don't have to. 

There is no concept of "current task" when we request new worker for a scheduling key. We just request one worker for this group. I can get the first task of the task_queue and just fail it (basically change the while to if), but doesn't seem to be much simpler?

Reply:
Got it, seems ok then.